### PR TITLE
drivers: counter: improved guard period doc

### LINKED ISSUE #57594: drivers: counter: improved documentation of guard period
**Describe the bug**

The current counter documentation (counter.h, line 115ff, struct counter_alarm_cfg) says:

> Absolute alarm cannot be set further in future than top_value
> decremented by the guard period. Relative alarm ticks cannot
> exceed current top value (see @ref counter_get_top_value).

And further down (line 555):

> Absolute value cannot exceed: (now + top_value - guard_period) % top_value.

The first documentation snippet is not wrong but confused me. Reading this for the first time I thought that the absolute tick value should not be set above top minus guard. It assumes that users understand the modulo operation in between.

The second documentation snippet is only correct when counting downwards (which several drivers don't). **When counting upwards, the given formula is wrong.**

**Proposed Change**

I propose something along the lines of:

> For both, relative and absolute alarms, the tick value can be anything between zero and the current top value (see @ref counter_get_top_value).
> 
> When setting an absolute value close to the current counter value there is a risk that the counter will have counted past the given absolute value before the driver manages to activate the alarm. Therefore a guard period can be defined that lets the driver decide whether it is late or not.
> 
> The guard period should be as many counter ticks as the driver will need at most to activate the timer after the driver API has been called. If the driver finds that the counter has just passed beyond the given absolute tick value but is still close enough to fall within the guard period, it will assume that it is "late", i.e. that the intended expiry time has already passed. Depending on the @ref COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE flag the driver will either ignore the alarm or expire it immediately in such a case.
> 
> If, however, the counter is past the given absolute tick value but outside the guard period, then the driver will assume that this is intentional and let the counter wrap around to/from zero before it expires.
> 
> More precisely:
> * When counting upwards (see @COUNTER_CONFIG_INFO_COUNT_UP) the given absolute tick value must be above (now + guard_period) % top_value
> * When counting downwards, the given absolute tick value must be less than (now + top_value - guard_period) % top_value.
> 
> Example:
> * counting upwards, now = 4950, top = 5000, guard period = 100, absolute ticks >= (4950 + 100) % 5000 = 50
> * counting downwards, now = 50, top = 5000, guard period = 100, absolute ticks <= (50 + 5000 - 100) % 5000 = 4950

If you agree, I can provide a PR with a text like this.

### PR DESCRIPTION
**Pull request by ghost**
The previous counter documentation said:

> Absolute alarm cannot be set further in future than top_value
> decremented by the guard period. Relative alarm ticks cannot
> exceed current top value (see @ref counter_get_top_value).

And further down:

> Absolute value cannot exceed: (now + top_value - guard_period)
> % top_value.

The first documentation snippet is not wrong but may be confusing. On casual reading one might assume that the absolute tick value should not be set above top minus guard. Users have to understand driver internals to follow that explanation.

The second documentation snippet is only correct when counting downwards (which several drivers don't). When counting upwards, the given formula is wrong.

This commit improves and corrects the documentation.

Fixes: #57594

### CODE REVIEW

**Review by henrikbrixandersen — DISMISSED** [review]
Thanks, looks good.

**Comment by fgrandel** [comment]
@nordic-krch ready for review

File: include/zephyr/drivers/counter.h [hunk 1]
```diff
@@ -545,22 +549,46 @@ static inline uint32_t z_impl_counter_get_top_value(const struct device *dev)
...
+ * Examples:
+ *
+ * - counting upwards, now = 4950, top value = 5000, guard period = 100:
+ *      absolute tick value >= (4950 + 100) % 5000 = 50
+ * - counting downwards, now = 50, top value = 5000, guard period = 100:
+ *      absolute tick value <= (50 + 5000 - * 100) % 5000 = 4950
+ *
+ * If you need only short alarm periods, you can set the guard period very high
+ * (e.g. half of the counter top value) which will make it highly unlikely that
+ * the counter will ever unintenionally wrap.
```
Comment (aasinclair) [inline include/zephyr/drivers/counter.h:589]: New description is very clear.

Nit: unintentionally

Comment (fgrandel) [reply include/zephyr/drivers/counter.h]: Good catch - fixed.

**Review by aasinclair — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**